staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -465,25 +482,91 @@ func (w *watchCache) waitUntilFreshAndBlock(resourceVersion uint64, trace *utilt
 }
 
+type ListResp struct {
+	objs    []interface{}
+	hasMore bool
+}
+
 // WaitUntilFreshAndList returns list of pointers to `storeElement` objects along
 // with their ResourceVersion and the name of the index, if any, that was used.
-func (w *watchCache) WaitUntilFreshAndList(resourceVersion uint64, matchValues []storage.MatchValue, trace *utiltrace.Trace) ([]interface{}, uint64, string, error) {
+func (w *watchCache) WaitUntilFreshAndList(resourceVersion uint64, key string, listOpts storage.ListOptions, matchValues []storage.MatchValue, trace *utiltrace.Trace) (ListResp, uint64, string, error) {
 	err := w.waitUntilFreshAndBlock(resourceVersion, trace)
 	defer w.RUnlock()
+
 	if err != nil {
-		return nil, 0, "", err
+		return ListResp{}, 0, "", err
+	}
+
+	if !strings.HasSuffix(key, "/") {
+		key += "/"
+	}
+
+	pred := listOpts.Predicate
+	hasContinuation := len(pred.Continue) > 0
+	var continueKey string
+
+	// Perform a clone under the lock, this should be a relatively
+	// inexpensive operation since the implementation of clone uses
+	// copy on write semantics. Once cloned, serve the list from the
+	// cloned copy to avoid building the response under a lock.
+	var storeClone btreeIndexer
+	if hasContinuation {
+		continueKey, _, err = storage.DecodeContinue(pred.Continue, key)
+		if err != nil {
+			return ListResp{}, 0, "", errors.NewBadRequest(fmt.Sprintf("invalid continue token: %v", err))
+		}
+		if !strings.HasSuffix(continueKey, "/") {
+			continueKey += "/"
+		}
+
+		if len(listOpts.ResourceVersion) > 0 && listOpts.ResourceVersion != "0" {
+			return ListResp{}, 0, "", errors.NewBadRequest("specifying resource version is not allowed when using continue")
+		}
+		if _, ok := w.continueCache.cache[resourceVersion]; !ok {
+			// We return a 410 Gone here for the following reason:
+			//
+			// Before the LIST request reaches the watchCache, we
+			// check if it should be delegated to etcd directly. In
+			// this check, we see if the request has a continuation,
+			// if it does, we check if the RV of the continuation
+			// token is still present in the watchCache or not, if
+			// it isn't then we let etcd serve the request.
+			//
+			// As and when events are removed from the watchCache
+			// (when it becomes full), we also check and evict the
+			// cached copy of the tree for the resource version whose
+			// event is going to be removed.
+			//
+			// Due to this, in case the cached clone is evicted, we
+			// return a 410 Gone similar to when a continue token
+			// expires. On receiving this error, the client can retry
+			// and on this retry, the check for delegation will route
+			// the request to etcd and things proceed accordingly.
+			err = errors.NewResourceExpired(fmt.Sprintf("too old resource version: %d", resourceVersion))
+			return ListResp{}, 0, "", err
+		}
+		storeClone = w.continueCache.cache[resourceVersion]
+	} else {
+		storeClone = w.store.Clone()
+		w.continueCache.cache[resourceVersion] = storeClone
```

Comment:
It needs a lock to avoid concurrenct map read/write, which crashes the kube-apiserver